internal/backends/mcu/renderer.rs

```diff
@@ -22,82 +21,130 @@ use i_slint_core::{Color, Coord, ImageInner, StaticTextures};
 
 type DirtyRegion = PhysicalRect;
 
-pub fn render_window_frame(
+pub trait LineBufferProvider {
+    /// Called before the frame is being drawn, with the dirty region. Return the actual dirty region
+    ///
+    /// The default implementation simply returns the dirty_region
+    fn set_dirty_region(&mut self, dirty_region: PhysicalRect) -> PhysicalRect {
+        dirty_region
+    }
+
+    /// Called once per line, you will have to call the render_fn back with the buffer.
+    fn process_line(
+        &mut self,
+        line: PhysicalLength,
+        render_fn: impl FnOnce(&mut [super::TargetPixel]),
+    );
+}
+
+#[derive(Default)]
+pub struct LineRenderer {
+    partial_cache: i_slint_core::item_rendering::PartialRenderingCache,
+}
+
+impl LineRenderer {
+    /// Render the window, line by line, into the buffer provided by the `line_buffer` function.
+    ///
+    /// The renderer uses a cache internally and will only render the part of the window
+    /// which are dirty. The `initial_dirty_region` is an extra dirty regin which will also
+    /// be rendered.
+    ///
+    /// TODO: the window should be the public slint::Window type
+    /// TODO: what about async and threading.
+    ///       (can we call the line_buffer function from different thread?)
+    /// TODO: should `initial_dirty_region` be set from a different call?
+    pub fn render(
+        &mut self,
+        window: Rc<i_slint_core::window::Window>,
+        initial_dirty_region: i_slint_core::item_rendering::DirtyRegion,
+        line_buffer: impl LineBufferProvider,
```

Comment:
I think it would be cleaner if this took a line_buffer by `&mut` and let the caller explicitly take call the code that's currently "hidden" in the `Drop` impl. What do you think?

Reply:
I actually considered doing it like that indeed. But i didn't do it because the user could also implement  `impl LineBufferProvider for &mut MyBuffer` if they wanted to.  I'm expecting these type to be quite small anyway (just a pointer to some buffer) and so making a pointer to a pointer seems counter productive.  But i actually don't know at this point how simple or complex the implementation will be for actual micro controlers.  Right know we piggy back on the Devices trait, but this will change.

I've implemented Drop in the current implementation because of the profiler.
I think we'll need to find a better way to do the profiling that doesn't need to get the devices to get the time.
Some devices even have profiling counter, so we could use the same trick as https://docs.rs/critical-section/latest/critical_section/macro.custom_impl.html using a #[no_mangle] symbol or something like that.

Reply:
Good point about implementing it for a reference type.